src/cluster/multi_step_cluster.cpp

```diff
@@ -20,105 +20,144 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+					r[c] = true;
+			return r;
+		}
+
+		vector<int> MultiStep::cluster(DatabaseFile& db, const vector<bool>* filter) {
+			statistics.reset();
+			config.command = Config::blastp;
+			//config.no_self_hits = true;
+			//config.output_format = { "6", "qnum", "snum" };
+			config.output_format = { "6", "qnum", "snum", "qcovhsp", "scovhsp", "bitscore" };
+			config.query_cover = 80;
+			config.subject_cover = 80;
+			config.algo = 0;
+			config.index_mode = 0;
+			config.freq_sd = 0;
+			config.max_alignments = numeric_limits<size_t>::max();
+
+			Workflow::Search::Options opt;
+			opt.db = &db;
+			opt.self = true;
+			Neighbors nb(db.ref_header.sequences);
+			opt.consumer = &nb;
+			opt.db_filter = filter;
+
+			Workflow::Search::run(opt);
+
+			message_stream << "Edges = " << nb.edges.size() << endl;
+
+			return Util::Algo::greedy_vortex_cover(nb);
+		}
+
+		vector <bool> MultiStep::steps(vector <vector<bool>> reps, vector <vector <int>> centroids) {
+			int count = reps.size() + 1;
+			int count_c = reps.size() - 1;
+			if (reps.empty()) { //first cluster step
+				vector <bool> rep_c(rep_bitset(centroids.front()));
+				size_t nrep = count_if(rep_c.begin(), rep_c.end(), [](bool x) {return x; });
+				message_stream << "Clustering step " << count << " complete. #Input sequences: " << centroids.front().size()<< " #Clusters: " << nrep << endl;
+				return rep_c;
+			}
+
+			else { //second, third,... cluster step defined
+				vector<bool> rep_c(rep_bitset(centroids.back(), &reps.back()));
+				for (size_t i = 0; i < centroids.back().size(); ++i)
+					if (!reps.back()[i])
+						centroids.back()[i] = centroids.back()[centroids[count_c][i]];
+
+				size_t n_rep_1 = count_if(reps.back().begin(), reps.back().end(), [](bool x) { return x; });
+				size_t n_rep_2 = count_if(rep_c.begin(), rep_c.end(), [](bool x) { return x; });
+
+				message_stream << "Clustering step " << count << " complete. #Input sequences: " << n_rep_1 << " #Clusters: " << n_rep_2 << endl;
+				return rep_c;
+			}
+		}
+		
+		void MultiStep::run() {
+			if (config.database == "")
+				throw runtime_error("Missing parameter: database file (--db/-d)");
+			config.command = Config::makedb;
+			unique_ptr<DatabaseFile> db(DatabaseFile::auto_create_from_fasta());
+			const size_t seq_count = db->ref_header.sequences;
+
+			vector<vector<int>> centroids{};
+			vector<vector<bool>> reps{};
+
+			vector<int> centroid1(cluster(*db, nullptr));
+			centroids.push_back(centroid1);
+			reps.push_back(steps(reps, centroids));
+
+			//no cluster steps defined
+			if (config.cluster_steps.empty()) {
+				config.sensitivity = Sensitivity::SENSITIVE;
+				vector<int> centroid_current(cluster(*db, &reps.back()));
+				centroids.push_back(centroid_current);
+				reps.push_back(steps(reps, centroids));
+			}
+			else { // cluster steps defined
+				for (int i = 0; i < config.cluster_steps.size(); i++) {
+					config.sensitivity = config.sens_map [config.cluster_steps[i]];
```

Comment:
Hier solltest du noch abfangen, ob der Benutzer in diesem Parameter ungültige Werte übergibt, und eine Ausnahme werfen.